File: api/src/routers/ai_keys.py
from typing import Optional
from fastapi import APIRouter, HTTPException, Depends
from pydantic import BaseModel, Field
import psycopg2.extras
import logging

from services.database.database import _get_conn, _put_conn, SafeId
from services.ai_client import mask_api_key, test_gemini_api_key, resolve_gemini_client
from routers.auth import get_current_user

logger = logging.getLogger("uvicorn.error")

router = APIRouter(tags=["Custom Integrated AI"])


class TestKeyRequest(BaseModel):
    api_key: str = Field(..., description="Google Gemini API key to test")


class SaveKeyRequest(BaseModel):
    api_key: str = Field(..., description="Google Gemini API key to persist")
    validate_first: Optional[bool] = Field(default=True, description="Whether to ping Gemini to ensure key is valid before saving")


# ----------------------------------------------------------------------
# 1. Test Key Verification Endpoint
# ----------------------------------------------------------------------
@router.post("/ai/test-key")
def test_key_endpoint(req: TestKeyRequest):
    """
    Test and verify whether an input Google Gemini API key is valid and operational.
    """
    res = test_gemini_api_key(req.api_key)
    return res


# ----------------------------------------------------------------------
# 2. Project-Level AI API Key Endpoints
# ----------------------------------------------------------------------
@router.get("/projects/{project_id}/ai-key")
def get_project_ai_key(project_id: SafeId):
    """
    Retrieve project custom AI key status and masked key representation.
    """
    conn = _get_conn()
    cur = None
    try:
        cur = conn.cursor(cursor_factory=psycopg2.extras.RealDictCursor)
        cur.execute("SELECT id, custom_ai_api_key FROM opentask.projects WHERE id = %s LIMIT 1;", (int(project_id),))
        row = cur.fetchone()
        if not row:
            raise HTTPException(status_code=404, detail="Project not found")

        key = row.get("custom_ai_api_key")
        has_custom = bool(key and key.strip())
        return {
            "project_id": str(project_id),
            "has_custom_key": has_custom,
            "masked_key": mask_api_key(key) if has_custom else None
        }
    finally:
        if cur is not None:
            cur.close()
        _put_conn(conn)


@router.post("/projects/{project_id}/ai-key")
def save_project_ai_key(project_id: SafeId, req: SaveKeyRequest):
    """
    Save or update custom Gemini AI API key for a project.
    """
    key = req.api_key.strip()
    if not key:
        raise HTTPException(status_code=400, detail="API key cannot be empty")

    if req.validate_first:
        validation = test_gemini_api_key(key)
        if not validation["valid"]:
            raise HTTPException(
                status_code=400,
                detail=f"Key validation failed: {validation.get('error') or 'Invalid API key'}"
            )

    conn = _get_conn()
    cur = None
    try:
        cur = conn.cursor(cursor_factory=psycopg2.extras.RealDictCursor)
        cur.execute(
            "UPDATE opentask.projects SET custom_ai_api_key = %s, updated_at = NOW() WHERE id = %s RETURNING id;",
            (key, int(project_id))
        )
        row = cur.fetchone()
        if not row:
            raise HTTPException(status_code=404, detail="Project not found")
        conn.commit()
        return {
            "project_id": str(project_id),
            "status": "saved",
            "has_custom_key": True,
            "masked_key": mask_api_key(key)
        }
    except HTTPException:
        conn.rollback()
        raise
    except Exception as e:
        conn.rollback()
        raise HTTPException(status_code=500, detail=str(e))
    finally:
        if cur is not None:
            cur.close()
        _put_conn(conn)


@router.delete("/projects/{project_id}/ai-key")
def delete_project_ai_key(project_id: SafeId):
    """
    Clear/remove custom Gemini AI API key for a project (reverting to system/user fallback).
    """
    conn = _get_conn()
    cur = None
    try:
        cur = conn.cursor(cursor_factory=psycopg2.extras.RealDictCursor)
        cur.execute(
            "UPDATE opentask.projects SET custom_ai_api_key = NULL, updated_at = NOW() WHERE id = %s RETURNING id;",
            (int(project_id),)
        )
        row = cur.fetchone()
        if not row:
            raise HTTPException(status_code=404, detail="Project not found")
        conn.commit()
        return {"project_id": str(project_id), "status": "removed", "has_custom_key": False}
    except HTTPException:
        conn.rollback()
        raise
    except Exception as e:
        conn.rollback()
        raise HTTPException(status_code=500, detail=str(e))
    finally:
        if cur is not None:
            cur.close()
        _put_conn(conn)


# ----------------------------------------------------------------------
# 3. User-Level Personal AI API Key Endpoints
# ----------------------------------------------------------------------
@router.get("/users/me/ai-key")
def get_user_ai_key(current_user: dict = Depends(get_current_user)):
    """
    Retrieve current authenticated user's personal AI key status.
    """
    user_id = current_user.get("id")
    if not user_id:
        raise HTTPException(status_code=401, detail="User not authenticated")

    conn = _get_conn()
    cur = None
    try:
        cur = conn.cursor(cursor_factory=psycopg2.extras.RealDictCursor)
        cur.execute("SELECT id, custom_ai_api_key FROM opentask.users WHERE id = %s LIMIT 1;", (int(user_id),))
        row = cur.fetchone()
        if not row:
            raise HTTPException(status_code=404, detail="User not found")

        key = row.get("custom_ai_api_key")
        has_custom = bool(key and key.strip())
        return {
            "user_id": str(user_id),
            "has_custom_key": has_custom,
            "masked_key": mask_api_key(key) if has_custom else None
        }
    finally:
        if cur is not None:
            cur.close()
        _put_conn(conn)


@router.post("/users/me/ai-key")
def save_user_ai_key(req: SaveKeyRequest, current_user: dict = Depends(get_current_user)):
    """
    Save or update personal Gemini AI API key for the current user.
    """
    user_id = current_user.get("id")
    if not user_id:
        raise HTTPException(status_code=401, detail="User not authenticated")

    key = req.api_key.strip()
    if not key:
        raise HTTPException(status_code=400, detail="API key cannot be empty")

    if req.validate_first:
        validation = test_gemini_api_key(key)
        if not validation["valid"]:
            raise HTTPException(
                status_code=400,
                detail=f"Key validation failed: {validation.get('error') or 'Invalid API key'}"
            )

    conn = _get_conn()
    cur = None
    try:
        cur = conn.cursor(cursor_factory=psycopg2.extras.RealDictCursor)
        cur.execute(
            "UPDATE opentask.users SET custom_ai_api_key = %s WHERE id = %s RETURNING id;",
            (key, int(user_id))
        )
        row = cur.fetchone()
        if not row:
            raise HTTPException(status_code=404, detail="User not found")
        conn.commit()
        return {
            "user_id": str(user_id),
            "status": "saved",
            "has_custom_key": True,
            "masked_key": mask_api_key(key)
        }
    except HTTPException:
        conn.rollback()
        raise
    except Exception as e:
        conn.rollback()
        raise HTTPException(status_code=500, detail=str(e))
    finally:
        if cur is not None:
            cur.close()
        _put_conn(conn)


@router.delete("/users/me/ai-key")
def delete_user_ai_key(current_user: dict = Depends(get_current_user)):
    """
    Clear/remove personal Gemini AI API key for the current user.
    """
    user_id = current_user.get("id")
    if not user_id:
        raise HTTPException(status_code=401, detail="User not authenticated")

    conn = _get_conn()
    cur = None
    try:
        cur = conn.cursor(cursor_factory=psycopg2.extras.RealDictCursor)
        cur.execute(
            "UPDATE opentask.users SET custom_ai_api_key = NULL WHERE id = %s RETURNING id;",
            (int(user_id),)
        )
        row = cur.fetchone()
        if not row:
            raise HTTPException(status_code=404, detail="User not found")
        conn.commit()
        return {"user_id": str(user_id), "status": "removed", "has_custom_key": False}
    except HTTPException:
        conn.rollback()
        raise
    except Exception as e:
        conn.rollback()
        raise HTTPException(status_code=500, detail=str(e))
    finally:
        if cur is not None:
            cur.close()
        _put_conn(conn)

File: opentask-api/src/routes/ai_keys.rs
use axum::{
    extract::{Path, State},
    http::StatusCode,
    routing::{get, post},
    Json, Router,
};
use sqlx::Row;

use crate::error::AppError;
use crate::extractors::auth::CurrentUser;
use crate::models::ai_key::{ApiKeyStatusResponse, SetApiKeyRequest, TestKeyRequest, TestKeyResponse};
use crate::models::safe_id::SafeId;
use crate::routes::auth::AppState;
use crate::services::gemini::{mask_api_key, test_gemini_api_key};

pub fn router() -> Router<AppState> {
    Router::new()
        .route("/ai/test-key", post(test_key))
        .route(
            "/projects/{project_id}/ai-key",
            get(get_project_ai_key).post(set_project_ai_key).delete(delete_project_ai_key),
        )
        .route(
            "/users/me/ai-key",
            get(get_user_ai_key).post(set_user_ai_key).delete(delete_user_ai_key),
        )
}

pub async fn test_key(Json(payload): Json<TestKeyRequest>) -> Json<TestKeyResponse> {
    let res = test_gemini_api_key(&payload.api_key).await;
    Json(res)
}

pub async fn get_project_ai_key(
    State(state): State<AppState>,
    Path(project_id): Path<SafeId>,
) -> Result<Json<ApiKeyStatusResponse>, AppError> {
    let row = sqlx::query("SELECT custom_ai_api_key FROM opentask.projects WHERE id = $1;")
        .bind(project_id.0)
        .fetch_optional(&state.pool)
        .await?;

    let key = row.and_then(|r| r.try_get::<Option<String>, _>("custom_ai_api_key").ok()).flatten();
    let masked = key.as_deref().and_then(mask_api_key);
    let has_key = masked.is_some();

    Ok(Json(ApiKeyStatusResponse {
        has_key,
        masked_key: masked,
        source: if has_key { "project".to_string() } else { "none".to_string() },
    }))
}

pub async fn set_project_ai_key(
    State(state): State<AppState>,
    Path(project_id): Path<SafeId>,
    Json(payload): Json<SetApiKeyRequest>,
) -> Result<Json<TestKeyResponse>, AppError> {
    let test_res = test_gemini_api_key(&payload.api_key).await;
    if !test_res.valid {
        return Err(AppError::BadRequest(
            test_res.error.unwrap_or_else(|| "Invalid Gemini API key".to_string()),
        ));
    }

    sqlx::query("UPDATE opentask.projects SET custom_ai_api_key = $1, updated_at = NOW() WHERE id = $2;")
        .bind(payload.api_key.trim())
        .bind(project_id.0)
        .execute(&state.pool)
        .await?;

    Ok(Json(test_res))
}

pub async fn delete_project_ai_key(
    State(state): State<AppState>,
    Path(project_id): Path<SafeId>,
) -> Result<StatusCode, AppError> {
    sqlx::query("UPDATE opentask.projects SET custom_ai_api_key = NULL, updated_at = NOW() WHERE id = $1;")
        .bind(project_id.0)
        .execute(&state.pool)
        .await?;

    Ok(StatusCode::NO_CONTENT)
}

pub async fn get_user_ai_key(
    State(state): State<AppState>,
    CurrentUser(user, _token): CurrentUser,
) -> Result<Json<ApiKeyStatusResponse>, AppError> {
    let row = sqlx::query("SELECT custom_ai_api_key FROM opentask.users WHERE gh_id = $1 OR gh_username = $2;")
        .bind(user.id.to_string())
        .bind(&user.login)
        .fetch_optional(&state.pool)
        .await?;

    let key = row.and_then(|r| r.try_get::<Option<String>, _>("custom_ai_api_key").ok()).flatten();
    let masked = key.as_deref().and_then(mask_api_key);
    let has_key = masked.is_some();

    Ok(Json(ApiKeyStatusResponse {
        has_key,
        masked_key: masked,
        source: if has_key { "user".to_string() } else { "none".to_string() },
    }))
}

pub async fn set_user_ai_key(
    State(state): State<AppState>,
    CurrentUser(user, _token): CurrentUser,
    Json(payload): Json<SetApiKeyRequest>,
) -> Result<Json<TestKeyResponse>, AppError> {
    let test_res = test_gemini_api_key(&payload.api_key).await;
    if !test_res.valid {
        return Err(AppError::BadRequest(
            test_res.error.unwrap_or_else(|| "Invalid Gemini API key".to_string()),
        ));
    }

    sqlx::query("UPDATE opentask.users SET custom_ai_api_key = $1 WHERE gh_id = $2 OR gh_username = $3;")
        .bind(payload.api_key.trim())
        .bind(user.id.to_string())
        .bind(&user.login)
        .execute(&state.pool)
        .await?;

    Ok(Json(test_res))
}

pub async fn delete_user_ai_key(
    State(state): State<AppState>,
    CurrentUser(user, _token): CurrentUser,
) -> Result<StatusCode, AppError> {
    sqlx::query("UPDATE opentask.users SET custom_ai_api_key = NULL WHERE gh_id = $1 OR gh_username = $2;")
        .bind(user.id.to_string())
        .bind(&user.login)
        .execute(&state.pool)
        .await?;

    Ok(StatusCode::NO_CONTENT)
}
